### ISSUE TITLE
Unable to build jasper-1.900.30

### ISSUE THREAD

**Issue by jridky** [mdadams/jasper#91]
Hi, 
unfortunately I have again problem with tests when I am trying to build the latest version of jasper.

```
FAIL: run_test_1
PASS: run_test_2
PASS: run_test_3
PASS: run_test_4
============================================================================
Testsuite summary for jasper 1.900.30
============================================================================
# TOTAL: 4
# PASS:  3
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See test/bin/test-suite.log
============================================================================

From run_test_1.log file:
############################################################
Testing invalid data sets
############################################################
############################################################
.....
############################################################
Input file: ./../../data/test/bad/00028-jasper-uaf-jas_realloc.jpg
Corrupt JPEG data: 19 extraneous bytes before marker 0xda
jpg 3 65281 33021 8 6466931703
actual exit status: 0
expected exit status: 1
ERROR: imginfo command had unexpected exit status  (expected 1 got 0)
############################################################
.....
############################################################
Input file: ./../../data/test/bad/1_crash.jpg
Corrupt JPEG data: 1 extraneous bytes before marker 0xc4
./run_test_1: line 53: 17999 Killed                  ASAN_OPTIONS="${special_asan_options[*]}" "$imginfo" "${imginfo_opts[@]}" < "$in_file"
actual exit status: 137
expected exit status: 1
ERROR: imginfo command had unexpected exit status  (expected 1 got 137)
############################################################
.....
############################################################
error count: 2
STATUS: FAILED
ERROR: FAILED
FAIL run_test_1 (exit status: 1)
```

In previous version (1.900.29), I have had problem with passing this test only on ARM architecture. Right now, I am unable to pass it even on x86-64 architecture.

**Comment by mdadams** [comment]
I think that this is just a re-surfacing of the old problem that different versions of the JPEG library behave differently for some of the JPEG test cases in the script run_test_1.  In other words, I think that the problem is simply with the test script not correctly accounting for different versions of the JPEG library.  Please try your test again with the code in the release tagged as version-1.900.31.  If you have any problems, let me know and I will re-open this issue.  If you do have problems though, please provide the output produced by running "imginfo --debug-level 1000" on  the failed JPEG file from run_test_1, as this would be helpful in diagnosing the problem.

**Comment by jridky** [comment]
Thank you for your update. At x86-64 architecture is now described problem resolved. Unfortunately at ARM architecture this issue still persist. Reason of failing run_test_1 is floating point exception (SIGFPE) in data/test/bad/00029-jasper-uninitvalue-jpc_pi_nextcprl.jp2 and data/test/bad/00052-jasper-misalignedaddress.jp2. Right now, I am waiting for imginfo output.  I will post it here, when it will be ready.

**Comment by jridky** [comment]
[bad-00059-jasper-misalignedaddress_jp2-log.txt](https://github.com/mdadams/jasper/files/606466/bad-00059-jasper-misalignedaddress_jp2-log.txt)
[bad-00029-jasper-uninitvalue-jpc_pi_nextcprl_jp2-log.txt](https://github.com/mdadams/jasper/files/606467/bad-00029-jasper-uninitvalue-jpc_pi_nextcprl_jp2-log.txt)

**Comment by mdadams** [comment]
As best I can tell, it seems that this might be due to an overflow problem that is benign on x86 but causes problems on ARM.  Would it be possible to get a backtrace of the imginfo program when it fails for each of these cases?  This would be helpful in isolating the problem.  Also, if you can provide the output of the program (from --debug-level 1000) with the address and undefined-behavior sanitizers enabled, this would be even better.  The undefined-behavior sanitizer would be helpful in catching overflow problems in the ARM case (assuming that this sanitizer also checks for overflow on ARM systems, as it does on x86).  Unfortunately, I don't have an ARM machine that I can use for testing.

**Comment by jridky** [comment]
Hi,
here you have backtraces from both cases.
[bad-00029-jasper-uninitvalue-jpc_pi_nextcprl_jp2-backtrace.txt](https://github.com/mdadams/jasper/files/608343/bad-00029-jasper-uninitvalue-jpc_pi_nextcprl_jp2-backtrace.txt)
[bad-00059-jasper-misalignedaddress_jp2-backtrace.txt](https://github.com/mdadams/jasper/files/608346/bad-00059-jasper-misalignedaddress_jp2-backtrace.txt)

Unfortunately, I am not so much familiar with debugging, so can you give me advice how to provide the output of the program with the address and undefined-behavior sanitizers enabled? I have never used it so I don't know how to provide that.

Thanks.

**Comment by mdadams** [comment]
Thank you for the backtraces.  They were helpful.  It looks like my suspicion of the source of the problem might be correct.  If my suspicion is correct, you should also see some overflow on integer shifts indicated by the undefined-behavior (UB) sanitizer sometime before the program crashes with a floating-point exception.  Therefore, it would be very helpful to have the program output with the sanitizers enabled.

No worries about not being familiar with the code sanitizers.  To enable one or more sanitizers, you just need to build JasPer with the desired sanitizers enabled and then run the resulting programs as you normally would.  If the extra sanitizer instrumentation added to the code (when it is buiilt with sanitizers enabled) detects anything wrong at run-time, an error/warning message will be output as part of the program execution.  The configure script for JasPer has options for enabling the address and UB sanitizers (for gcc- and clang-based builds) as well as some other sanitizers.  It you use "--enable-asan" and "--enable-usan", this will enable both the address and UB sanitizers, respectively, when you build JasPer.  So, just configure JasPer with these options added to whatever options you normally use for configure and then collect the program output as you did before.  Incidentally, which version of GCC (or Clang) are you using?

**Comment by jridky** [comment]
Thanks a lot for advice.
Here you have outputs from imginfo --debug-level 1000 command.

[bad-00029-uninitvalue-jpc_pi_nextcprl-sanitylog.txt](https://github.com/mdadams/jasper/files/610640/bad-00029-uninitvalue-jpc_pi_nextcprl-sanitylog.txt)
[bad-00052-misalignedaddress-sanitylog.txt](https://github.com/mdadams/jasper/files/610641/bad-00052-misalignedaddress-sanitylog.txt)

**Comment by mdadams** [comment]
@jridky Thank you for the above files.  They allowed me to verify that the problem was what I thought it to be.  I think that the problem should now be fixed in the new version of the code with the tag version-2.0.0-beta.2.  (The main fix went into commit aa0b0f79ade5eef8b0e7a214c03f5af54b36ba7d.)  Please note that, as of version 2 of JasPer, the build is performed with cmake (not autotools).  The INSTALL file has some revised instructions on how to build the code (with cmake) that might be helpful.
